pkg/resources/sku/sku.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func GetSKU(SKUId string, SKUConfig *corev1.ConfigMap, retSku *SKU, isUpdated bool) error {
+	allSKUs := &[]skuConfigReceiver{}
+	err := json.Unmarshal([]byte(SKUConfig.Data["sku-configs"]), allSKUs)
+	if err != nil {
+		return err
+	}
+	skuReceiver := skuConfigReceiver{}
+	for _, sku := range *allSKUs {
+		if sku.Name == SKUId {
+			skuReceiver = sku
+			break
+		}
+	}
+	if skuReceiver.Name == "" {
+		return errors.New(fmt.Sprintf("could not find sku config with name '%s'", SKUId))
+	}
+
+	// map of products iterate over that to build the return map
+	products := map[v1alpha1.ProductName][]string{
+		v1alpha1.Product3Scale: {
+			"backend_listener",
+			"backend_worker",
+			"apicast_production",
+		},
+		v1alpha1.ProductRHSSOUser: {
+			"keycloak",
+		},
+		v1alpha1.ProductMarin3r: {
+			"marin3r",
+		},
+	}
+	retSku.name = SKUId
+	retSku.productConfigs = map[v1alpha1.ProductName]ProductConfig{}
+	retSku.isUpdated = isUpdated
+
+	// loop through array of ddcss (deployment deploymentConfig StatefulSets)
+	for product, ddcssNames := range products {
+		pc := ProductConfig{
+			sku:             retSku,
+			productName:     product,
+			resourceConfigs: map[string]ResourceConfig{},
+		}
+		// if any are missing set sane defaults
+		for _, ddcssName := range ddcssNames {
+			pc.resourceConfigs[ddcssName] = skuReceiver.Resources[ddcssName]
+		}
+		retSku.productConfigs[product] = pc
+	}
+	return nil
+}
+
+func (s *SKU) GetProduct(productName v1alpha1.ProductName) ProductConfig {
+	return s.productConfigs[productName]
+}
+
+func (s *SKU) IsUpdated() bool {
+	return s.isUpdated
+}
+
+func (p *ProductConfig) GetResourceConfig(ddcssName string) corev1.ResourceRequirements {
+	return p.resourceConfigs[ddcssName].Resources
+}
+
+func (p ProductConfig) GetReplicas(ddcssName string) int32 {
+	return p.resourceConfigs[ddcssName].Replicas
+}
+
+func (p *ProductConfig) Configure(obj interface{}) error {
+	switch t := obj.(type) {
+	case *appsv1.DeploymentConfig:
+		if p.sku.isUpdated && t.Spec.Replicas < p.resourceConfigs[t.ObjectMeta.Name].Replicas {
+			t.Spec.Replicas = p.resourceConfigs[t.ObjectMeta.Name].Replicas
+		}
+		p.mutate(t.Spec.Template, t.ObjectMeta.Name)
+		break
+	case *appsv12.Deployment:
+		configReplicas := p.resourceConfigs[t.ObjectMeta.Name].Replicas
+		if p.sku.isUpdated && *t.Spec.Replicas < p.resourceConfigs[t.ObjectMeta.Name].Replicas {
+			t.Spec.Replicas = &configReplicas
+		}
+		p.mutate(&t.Spec.Template, t.ObjectMeta.Name)
+		break
+	case *appsv12.StatefulSet:
+		configReplicas := p.resourceConfigs[t.ObjectMeta.Name].Replicas
+		if p.sku.isUpdated && *t.Spec.Replicas < p.resourceConfigs[t.ObjectMeta.Name].Replicas {
+			t.Spec.Replicas = &configReplicas
+		}
+		p.mutate(&t.Spec.Template, t.ObjectMeta.Name)
+		break
+	default:
+		return errors.New(fmt.Sprintf("sku configuration can only be applied to Deployments, StatefulSets or Deployment Configs, found %s", reflect.TypeOf(obj)))
+	}
+	return nil
+}
```

Comment:
Hey @sergioifg94 I'm implemented some of the feedback here and will push the change soon. The one piece I didn't take was that the function would accept a Object rather than an interface. the reason for this is that we would also like this function to handle APIManager crs in the case of 3scale so there will be some refactoring later on this.

Reply:
Thanks @laurafitzgerald. I think it would still handle APIManager CRs. The `metav1.Object` is implemented by every resource, including Custom Resources

Reply:
ah didn't know that. I will make the update. 👍